packages/core/src/new-api/internal/validation/validateStageTwo.ts

```diff
@@ -19,105 +25,122 @@ export async function validateStageTwo(
   artifactLoader: ArtifactResolver,
   deploymentParameters: DeploymentParameters,
   accounts: string[]
-): Promise<void> {
+): Promise<ValidationErrorDeploymentResult | null> {
   const futures = getFuturesFromModule(module);
 
-  // originally, I wrote a getSubmodulesFromModule function similar to the one above
-  // that recursively retrieved all submodules regardless of how deeply nested they were.
-  // however, by taking only the top level submodules of the current depth and recursively
-  // validating each of those, we achieve the same effect.
-  const submodules = module.submodules;
-  for (const submodule of submodules) {
```

Comment:
Get futures from module, gets the futures of all submodules as well recursively. Have I missed a case that submodule recursion gets us here?

Reply:
Sounds right! We should have a small test for this though

Reply:
Whoever does the validation multi-result work can add tests around this as well.